### ISSUE TITLE
Work as semaphore or as mutex, not mixed it

### ISSUE THREAD

**Issue by Fapfood** [cloudfoundry-community/locker#2]
Now when you lock, you increment counter (as semaphore) but when unlock you assign 0 (as mutex). I think it should either prevent from lock multiple times (as mutex) or force to unlock as many times as lock (as semaphore). Due to backward compatibility, you can implement unlock (which works like now) and e.g. downlock (which can decrement counter) in locker resourse.

**Comment by jhunt** [comment]
@Fapfood can you point to specific code where you see the unlock operation assigning the incremented counter to 0?

**Comment by Fapfood** [comment]
Here increment:
https://github.com/cloudfoundry-community/locker/blob/5b003cdbc39ef7e27d4cc62251573dcfe9b83cc5/locker.go#L77

Here assign 0:
https://github.com/cloudfoundry-community/locker/blob/5b003cdbc39ef7e27d4cc62251573dcfe9b83cc5/locker.go#L93

Problem with this is when you trigger pipeline multiple times they will have same key/locked_by and will increment counter, but when first running of pipeline unlock the lock, the full lock will be released. Other runs of pipeline should still have this lock locked.

**Comment by jhunt** [comment]
Are you not running this with serial jobs in Concourse?

**Comment by Fapfood** [comment]
Yes, but this doesn't solve problem in each cases. Consider pipeline with more than one job. You acquire lock before the first job and release it after the last. Serial prevent from running first job twice in the same time, but you can run first and second job at the same time. serial_groups also dosn't solve this problem, because start job in random order.
